target/linux/ramips/patches-5.4/202-mt7621-pci-re-do-reset-boot-process.patch

```diff
@@ -0,0 +1,224 @@
+
+Try to fix that is the main motivation of this patch series.
+
+Also in openwrt there is a driver for mt7621-pci which seems was rewritten
+from scratch (for kernel 4.14) by Ryder Lee and Weijie Gao from mediatek. 
+There the approach for reset assert-deassert process is to set as 'gpio'
+the function for all the 'pcie' group for the pinctrl driver and use those
+gpio's as a reset for the end points. The driver I am talking about is still
+using legacy pci and legacy gpio kernel interfaces. IMHO, the correct thing
+to do is make this staging driver properly clean and functional and put it
+in its correct place in the mainline.
+
+See:
+* https://gist.github.com/dengqf6/7a9e9b4032d99f1a91dd9256c8a65c36
+
+Because of all of this this patch series tries to avoid random hangs of boot
+trying to use the 'reset-gpios' approach.
+
+Changes are only compile tested.
+
+Hope this helps.
+
+Best regards,
+    Sergio Paracuellos
+
+--- a/drivers/staging/mt7621-pci/pci-mt7621.c
++++ b/drivers/staging/mt7621-pci/pci-mt7621.c
+@@ -35,9 +35,6 @@
+ 
+ #include "../../pci/pci.h"
+ 
+-/* sysctl */
+-#define MT7621_GPIO_MODE		0x60
+-
+ /* MediaTek specific configuration registers */
+ #define PCIE_FTS_NUM			0x70c
+ #define PCIE_FTS_NUM_MASK		GENMASK(15, 8)
+@@ -90,9 +87,7 @@
+ #define PCIE_CLK_GEN1_DIS		GENMASK(30, 24)
+ #define PCIE_CLK_GEN1_EN		(BIT(27) | BIT(25))
+ #define MEMORY_BASE			0x0
+-#define PERST_MODE_MASK			GENMASK(11, 10)
+-#define PERST_MODE_GPIO			BIT(10)
+-#define PERST_DELAY_US			1000
++#define PERST_DELAY_US			100
+ 
+ /**
+  * struct mt7621_pcie_port - PCIe port information
+@@ -101,6 +96,7 @@
+  * @pcie: pointer to PCIe host info
+  * @phy: pointer to PHY control block
+  * @pcie_rst: pointer to port reset control
++ * @gpio_rst: gpio reset
+  * @slot: port slot
+  * @enabled: indicates if port is enabled
+  */
+@@ -110,6 +106,7 @@ struct mt7621_pcie_port {
+ 	struct mt7621_pcie *pcie;
+ 	struct phy *phy;
+ 	struct reset_control *pcie_rst;
++	struct gpio_desc *gpio_rst;
+ 	u32 slot;
+ 	bool enabled;
+ };
+@@ -123,8 +120,6 @@ struct mt7621_pcie_port {
+  * @offset: IO / Memory offset
+  * @dev: Pointer to PCIe device
+  * @ports: pointer to PCIe port information
+- * @perst: gpio reset
+- * @rst: pointer to pcie reset
+  * @resets_inverted: depends on chip revision
+  * reset lines are inverted.
+  */
+@@ -139,8 +134,6 @@ struct mt7621_pcie {
+ 		resource_size_t io;
+ 	} offset;
+ 	struct list_head ports;
+-	struct gpio_desc *perst;
+-	struct reset_control *rst;
+ 	bool resets_inverted;
+ };
+ 
+@@ -207,16 +200,14 @@ static void write_config(struct mt7621_p
+ 	pcie_write(pcie, val, RALINK_PCI_CONFIG_DATA);
+ }
+ 
+-static inline void mt7621_perst_gpio_pcie_assert(struct mt7621_pcie *pcie)
++static inline void mt7621_rst_gpio_pcie_assert(struct mt7621_pcie_port *port)
+ {
+-	gpiod_set_value(pcie->perst, 0);
+-	mdelay(PERST_DELAY_US);
++	gpiod_set_value(port->gpio_rst, 0);
```

Comment:
@paraka gpiod_set_value will invert the value if it's defined as ACTIVE_LOW in dts. My assumption is that you now have all pcie devices in reset and you could check by:
```bash
cat /sys/kernel/debug/gpio
```

Reply:
and upstream gpio driver may be buggy according to @vDorst in previous comment.

Reply:
You're right but that's the first thing I checked, and tried fixing it using GPIO_ACTIVE_HIGH in dts but no luck

Reply:
@dengqf6  Can you try to avoid the FTS_NUM read and see what happens?

diff --git a/drivers/staging/mt7621-pci/pci-mt7621.c b/drivers/staging/mt7621-pci/pci-mt7621.c
index fb6e65f7624b..7d49f350e96f 100644
--- a/drivers/staging/mt7621-pci/pci-mt7621.c
+++ b/drivers/staging/mt7621-pci/pci-mt7621.c
@@ -494,10 +494,6 @@ static void mt7621_pcie_init_ports(struct mt7621_pcie *pcie)
                if (err) {
                        dev_err(dev, "Initiating port %d failed\n", slot);
                        list_del(&port->list);
-               } else {
-                       val = read_config(pcie, slot, PCIE_FTS_NUM);
-                       dev_info(dev, "Port %d N_FTS = %x\n", slot,
-                                (unsigned int)val);
                }
        }

Reply:
```
[    4.275998] mt7621-pci-phy 1e149000.pcie-phy: Xtal is 40MHz
[    4.395982] mt7621-pci-phy 1e149000.pcie-phy: Xtal is 40MHz
[    4.515984] mt7621-pci-phy 1e14a000.pcie-phy: Xtal is 40MHz
[    4.621476] mt7621-pci 1e140000.pcie: pcie0 no card, disable it (RST & CLK)
[    4.628429] mt7621-pci 1e140000.pcie: pcie1 no card, disable it (RST & CLK)
[    4.635359] mt7621-pci 1e140000.pcie: pcie2 no card, disable it (RST & CLK)
[    4.642309] mt7621-pci 1e140000.pcie: Nothing is connected in virtual bridges. Exiting...
```

Reply:
So it is not hanging anymore, right? The phy seems to be ok but the endpoints have no link at all... It seems GPIO MODE is not set at all and is needed (which is weird because the driver for 4.14 you upload the other day does not set it).

Try this patch please:

```
diff --git a/drivers/staging/mt7621-pci/pci-mt7621.c b/drivers/staging/mt7621-pci/pci-mt7621.c
index fb6e65f7624b..0880a21f2620 100644
--- a/drivers/staging/mt7621-pci/pci-mt7621.c
+++ b/drivers/staging/mt7621-pci/pci-mt7621.c
@@ -35,6 +35,9 @@
 
 #include "../../pci/pci.h"
 
+/* sysctl */
+#define MT7621_GPIO_MODE               0x60
+
 /* MediaTek specific configuration registers */
 #define PCIE_FTS_NUM                   0x70c
 #define PCIE_FTS_NUM_MASK              GENMASK(15, 8)
@@ -81,6 +84,8 @@
 #define PCIE_PORT_LINKUP               BIT(0)
 
 #define MEMORY_BASE                    0x0
+#define PERST_MODE_MASK                        GENMASK(11, 10)
+#define PERST_MODE_GPIO                        BIT(10)
 #define PERST_DELAY_US                 100
 
 /**
@@ -481,9 +486,10 @@ static void mt7621_pcie_init_ports(struct mt7621_pcie *pcie)
 {
        struct device *dev = pcie->dev;
        struct mt7621_pcie_port *port, *tmp;
-       u32 val = 0;
        int err;
 
+       rt_sysc_m32(PERST_MODE_MASK, PERST_MODE_GPIO, MT7621_GPIO_MODE);
+
        mt7621_pcie_reset_assert(pcie);
        mt7621_pcie_reset_rc_deassert(pcie);
 
@@ -494,10 +500,6 @@ static void mt7621_pcie_init_ports(struct mt7621_pcie *pcie)
                if (err) {
                        dev_err(dev, "Initiating port %d failed\n", slot);
                        list_del(&port->list);
-               } else {
-                       val = read_config(pcie, slot, PCIE_FTS_NUM);
-                       dev_info(dev, "Port %d N_FTS = %x\n", slot,
-                                (unsigned int)val);
                }
        }
 ```

Reply:
> So it is not hanging anymore, right? The phy seems to be ok but the endpoints have no link at all... It seems GPIO MODE is not set at all and is needed (which is weird because the driver for 4.14 you upload the other day does not set it).

That comes from the pinctrl property: staging pinctrl driver uses "group" and "function" instead of "ralink,group" and "ralink,function" for pinctrl nodes. It seems that this change is missing in the PR here.

Reply:
> That comes from the pinctrl property: staging pinctrl driver uses "group" and "function" instead of "ralink,group" and "ralink,function" for pinctrl nodes. It seems that this change is missing in the PR here.

@981213 That's because ralink,function and ralink,group are not standard properties and group and function are. All the nodes in mt7621-dts in staging kernel are using those:

https://git.kernel.org/pub/scm/linux/kernel/git/gregkh/staging.git/tree/drivers/staging/mt7621-dts/mt7621.dtsi?h=staging-testing

Reply:
I'm aware of this commit and I'm just pointing out a missing dts change, which should fix the problem you mentioned earlier :D

Reply:
It seems to work now, I will try rebooting a few more times to check

Another problem: R3G's LEDs fail to initialize because one of them uses GPIO8
You should release the GPIO resource after PCIe initialization

Reply:
@dengqf6 Would you mind to test whether fixing the dts property mentioned above solves the problem as well? PCIe driver is a wrong place to setup pinctrl in the first place.

Reply:
@dengqf6 Good to know :-)

Try this patch for the LED issue:

```
diff --git a/drivers/staging/mt7621-pci/pci-mt7621.c b/drivers/staging/mt7621-pci/pci-mt7621.c
index 0880a21f2620..8399e4629e38 100644
--- a/drivers/staging/mt7621-pci/pci-mt7621.c
+++ b/drivers/staging/mt7621-pci/pci-mt7621.c
@@ -482,6 +482,14 @@ static void mt7621_pcie_reset_ep_deassert(struct mt7621_pcie *pcie)
        mdelay(PERST_DELAY_US);
 }
 
+static void mt7621_pcie_release_gpios(struct mt7621_pcie *pcie)
+{
+       struct mt7621_pcie_port *port;
+
+       list_for_each_entry(port, &pcie->ports, list)
+               gpiod_put(port->gpio_rst);
+}
+
 static void mt7621_pcie_init_ports(struct mt7621_pcie *pcie)
 {
        struct device *dev = pcie->dev;
@@ -706,6 +714,8 @@ static int mt7621_pci_probe(struct platform_device *pdev)
                return err;
        }
 
+       mt7621_pcie_release_gpios(pcie);
+
        return 0;
 }
```

Reply:
Update: reboots were fine (tried 10 times in a row), but a power cycle can still cause PCIe to not detect cards

Reply:
@paraka I just checked schematics I have for two mt7621 boards, One of it has both PCIe wifi cards connected together to the same PERST_N pin. The other connected two wifi cards together to PERST_N and the 3rd PCIe reset is connected to a GPIO. I think this should be taken into account. We should only occupy PERST_N pin by default and 2nd/3rd resets should be added on-demand per-device.

Reply:
@dengqf6  Do you mean a timing issue?

Reply:
@981213 Do you mean define gpio 19 as perst-gpio to be mandatory and reset-gpios the other two making them optional?

Reply:
GPIO is broken after I replaced `ralink,group` with `groups` and `ralink,function` with `function`
```
[    4.048051] rt2880-pinmux pinctrl: pcie is already enabled
[    4.053518] mt7621-pci 1e140000.pcie: Error applying setting, reverse things back
```

Reply:
After a few more tests, it seems that avoiding the `FTS_NUM` read can mostly workaround this issue (at least for soft reboot)

Reply:
All this changes in new v2 of the series:

http://driverdev.linuxdriverproject.org/pipermail/driverdev-devel/2020-March/142338.html

Reply:
@dengqf6 

> GPIO is broken after I replaced ralink,group with groups and ralink,function with function

One more driver to fix then :( I'll try to figure this one out.

> Do you mean define gpio 19 as perst-gpio to be mandatory and reset-gpios the other two making them optional?

Exactly. And by requesting only pins used, formerly mentioned R3G problem can be solved without releasing GPIO resources.
BTW you forgot to fix the formerly mentioned inverted reset line issue in your v2.

Reply:
@981213 

I am a bit lost. Current 4.14 device tree set to GPIO_ACTIVE_LOW this gpio pins in the dts ... Why is now necessary to set up as GPIO_ACTIVE_HIGH?

Reply:
@paraka

> Current 4.14 device tree set to GPIO_ACTIVE_LOW this gpio pins in the dts

Early gpio apis like gpio_set_value takes only gpio number from device tree and set pin voltage level directly. This means we lost the ACTIVE_HIGH/ACTIVE_LOW info from dts. The later introduced gpiod apis take this property into account and invert value if the pin is configured as active low.

> Why is now necessary to set up as GPIO_ACTIVE_HIGH?

pcie reset is technically active low so I think the proper way is to invert the value in driver instead of configuring gpio pins as active high.

Reply:
@981213 

Do you mean change value when requested? 

From:
port->gpio_rst = devm_gpiod_get_index(dev, "reset", slot, GPIOD_OUT_HIGH);
To:
port->gpio_rst = devm_gpiod_get_index(dev, "reset", slot, GPIOD_OUT_LOW);

Reply:
@paraka

> Do you mean change value when requested?
> 
> From:
> port->gpio_rst = devm_gpiod_get_index(dev, "reset", slot, GPIOD_OUT_HIGH);
> To:
> port->gpio_rst = devm_gpiod_get_index(dev, "reset", slot, GPIOD_OUT_LOW);

This one as well as gpiod_set_value calls in mt7621_rst_gpio_pcie_assert/deassert

Reply:
Ok, thanks, I will do this changes and also avoid to fail on parsing the dt if gpio cannot be requested to handle the board dependant issue. Will send link with v3 in a while.

Reply:
V3 of this series already sent: http://driverdev.linuxdriverproject.org/pipermail/driverdev-devel/2020-March/142357.html